blog/_posts/2021-05-26-wip-vs-harness.md

```diff
@@ -0,0 +1,138 @@
+conclude: these are mutually exclusive approaches. Some projects are WIP
+projects. And some projects are verbose projects.
+
+Maybe that analogy is illuminating. But if it's not, I'm going to go into more
+detail on some of the patterns we've seen on:
+
+- What problems GitOps & local Kubernetes dev solve
+
+- How these two toolsets approach those problems
+
+- How to bridge the gap between them
+
+## More Service, More Problems
+
+If you're working on a multi-service app, you have two types of changes:
+
+- WIP changes - No services are running in dev, and you want to make a change.
+
+- Harness changes - All your services are running in prod, and you want to make a change.
+
+These two sets may sound the same but they are mostly *non-overlapping*!
+
+Here's a table to help illustrate the differences:
+
+|          | WIP Changes | Harness Changes |
+----------------|-------------|------------------
+| Add segfaults to see what breaks                 | ✅ | ❌️                             |
+| Bring up all services from scratch               | ✅ | Something terrible has happened |
+| Reset the database                               | ✅️ | ❌ ❌ ❌                      |
+| Sacifice correctness for faster updates | ✅  |     ❌️                            |
+| Add security backdoors for testing | ✅ | 🚑 |
+| Audit records of every change | ❌️| ✅ |
+| Only one update at a time | ❌️| ✅ |
+| Rolling restarts | ❌ | ✅ |
+| Rollbacks | 🤷 | 🙏 |
+| Automatic reverts if too many errors | ❌ | ✅ |
+| Immutable containers for security | ❌ | ✅ |
+
+GitOps (ArgoCD, CodeFresh) is an approach to Harness development:
+
+- A service in the cloud listens to changes to your Git repo.
+
+- When your Git repo changes, the GitOps server updates your prod env to match
+  what's in your Git repo.
+
+- Git serves as the system of record for changes to your prod env.
+
+Local K8s (Tilt, Skaffold) is an approach to WIP development:
+
+- A service locally listens to changes to your filesystem.
+
+- When your file system changes, the Tilt server updates your dev env to match
+  what's on disk.
+
+- Tilt doesn't worry to much about recording the history of changes, assuming
+  that you can throw it all away when you're done.
+
+## Blurred Lines
+
+We recommend you use both GitOps and local K8s together! Tilt-team does this.
+
+For local dev and CI, we use Tilt to run services against [a one-time-use KIND
+cluster](https://blog.tilt.dev/2021/04/02/kubernetes-on-ci.html).
+
+For production releases, we push a tag to our GitHub repo, and the CI system
+will automatically create release binaries and update services.
+
+We even work with teams who use Tilt to run ArgoCD in dev, to make changes to
+how their GitOps pipeline works together.
+
+But there are teams who mix and match these tools in more creative ways!
+
+---
+
+### Local K8s for Harness Development
+
+- I once met a team who needed to deploy to an on-prem mainframe. They created
```

Comment:
What is this paragraph getting at?

Reply:
there ARE teams who use "tools for WIP/pre-commit dev" to deploy their production stuff. it's not...wrong, just a weird use of that tooling.

In the early NodeJS days, there were major problems where people would serve the `.git` directories of their code in production, and accidentally leak their whole git history :grimacing:

Reply:
i added some context above